Prepend `SyntaxStringInterpolationError` errors in macro expansion with `Internal macro error:`

natikgadzhi opened the pull request:
## Summary
This pull request prepends `SyntaxStringInterpolationError` message with `Internal macro error:`. Closes #2039.

## Changes
- In `MacroExpansionContext.addDiagnostics`, if the passed `error as? SyntaxStringInterpolationError` worked, we prepend it with the `Internal macro error` and grab it's description. Otherwise, we just use `String(describing:error)` as the diagnostic message.
- Adds a basic test for `addDiagnostics` for that behavior.

## Next steps
- @DougGregor, does this look like the right direction to you? I'm new to Macros, and I'm here to learn — no hard feelings if my attempt is naive / plain wrong.
- @Matejkob, sounds like your area of interest! Could you give me some hints to how I could test this end to end, and construct a test macro that would throw syntax string interpolation error?

natikgadzhi commented on Sources/SwiftSyntaxBuilder/Syntax+StringInterpolation.swift:
```diff
@@ -161,15 +161,18 @@ where Self.StringInterpolation == SyntaxStringInterpolation {
+// Used in `MacroExpansionContext` to make a nicer error message when a
+// macro expansion fails with SwiftSyntaxBuilder making a syntax node
+// with string interpolation.
+@_spi(Diagnostics) public enum SyntaxStringInterpolationError: Error, CustomStringConvertible {
```
Don't know if it's okay to `@_spi` this, but that seemed like a better option than to make it public outright.

ahoppen replied:
I just checked and `SyntaxStringInterpolationError` currently serves two distinct use cases: 
1. The `.diagnostics` case in `SyntaxProtocol(validating:)`
2. The `.producedInvalidNodeType` case in SyntaxNodeWithBody and `DeclSyntaxProtocol.init`

I think what we should do, is split the error into two distinct structs and make the one that represents the `producedInvalidNodeType` case public. That way, we minimize the public surface are we are introducing while also allowing other clients to catch that case.

natikgadzhi commented on Sources/SwiftSyntaxBuilder/Syntax+StringInterpolation.swift:
```diff
@@ -161,15 +161,18 @@ where Self.StringInterpolation == SyntaxStringInterpolation {
   }
 
-  var description: String {
+  public var description: String {
```
I _think_ compiler complained about this for custom string convertible.

natikgadzhi commented on Tests/SwiftSyntaxMacroExpansionTest/MacroExpansionContextTests.swift:
```diff
@@ -0,0 +1,73 @@
+import SwiftSyntaxMacrosTestSupport
+import XCTest
+
+//private class TestMacroExpansionContext: MacroExpansionContext {
```
TODO: Remove the commented out test expansion context, basic context worked out fine.

ahoppen commented on Tests/SwiftSyntaxMacroExpansionTest/MacroExpansionContextTests.swift:
```diff
@@ -0,0 +1,73 @@
+final class MacroExpansionContextTests: XCTestCase {
+
+  func testMacroExpansionContextAddDiagnosticsAddsSwiftSyntaxInterpolationErrorsWithWrappingMessage(){
+    let context = BasicMacroExpansionContext()
+    let error = SyntaxStringInterpolationError.producedInvalidNodeType(expectedType: DeclSyntax.self, actualType: ExprSyntax.self)
+    // Since we only care about the error switch inside of addDagnostics, we don't care about the particular node we're passing in.
+    context.addDiagnostics(from: error, node: ExprSyntax("1"))
+
+    XCTAssertEqual(context.diagnostics.count, 1)
+    XCTAssertTrue(context.diagnostics.first!.message.starts(with: "Internal macro error:"))
```
I think you should be able to test the real-case scenario when this happens by defining an extension macro that returns 

```swift
try ExtensionDeclSyntax("var x: Int")
```

And then use `assertMacroExpansion` to check that the emitted error contains `Internal macro error:`

natikgadzhi commented:
@ahoppen, thank you for the review! 

I've cleaned up the code and added an end-to-end test for now. The next part is splitting the error enum into two pieces. That'll be a separate commit.

natikgadzhi commented:
@ahoppen ready for review!

ahoppen commented on Sources/SwiftSyntaxBuilder/Syntax+StringInterpolation.swift:
```diff
@@ -161,23 +161,28 @@ where Self.StringInterpolation == SyntaxStringInterpolation {
-    case .producedInvalidNodeType(expectedType: let expectedType, actualType: let actualType):
-      return "Parsing the code snippet was expected to produce a \(expectedType) but produced a \(actualType)"
-    case .diagnostics(let diagnostics, let tree):
-      // Start the diagnostic on a new line so it isn't prefixed with the file, which messes up the
-      // column-aligned message from ``DiagnosticsFormatter``.
```
Could we keep this comment?

ahoppen commented on Sources/SwiftSyntaxMacros/MacroExpansionContext.swift:
```diff
@@ -12,6 +12,7 @@
 
 import SwiftDiagnostics
 import SwiftSyntax
+@_spi(Diagnostics) import SwiftSyntaxBuilder
```
I don’t think we need `@_spi(Diagnostics)` anymore.

natikgadzhi replied:
Good catch! Sorry I missed it.

ahoppen commented on Tests/SwiftSyntaxMacroExpansionTest/MacroExpansionContextTests.swift:
```diff
@@ -0,0 +1,79 @@
+  }
+}
+
+final class MacroExpansionContextTests: XCTestCase {
```
I think a better name for this test would now be something like `StringInterpolationErrorTests`.

ahoppen reviewed (commented):
A couple of minor comments, otherwise LGTM.

natikgadzhi commented:
Oooops, those are embarrassing. I’ll clean it up.

natikgadzhi commented:
@ahoppen done!

ahoppen commented on Tests/SwiftSyntaxMacroExpansionTest/StringInterpolationErrorTests.swift:
```diff
@@ -0,0 +1,79 @@
+    // Since we only care about the error switch inside of addDagnostics, we don't care about the particular node we're passing in
+    context.addDiagnostics(from: error, node: ExprSyntax("1"))
+    XCTAssertEqual(context.diagnostics.count, 1)
+    XCTAssertTrue(context.diagnostics.first!.message.starts(with: "Internal macro error:"))
```
Can you do the following? That way the test doesn’t crash if there is no diagnostic.

```suggestion
    let diagnostic = try XCTUnwrap(context.diagnostics.first)
    XCTAssertTrue(diagnostic.message.starts(with: "Internal macro error:"))
```

Same for the test below

natikgadzhi replied:
Yep, good call. Fixed.

ahoppen reviewed (commented):
One minor comment, otherwise LGTM.

ahoppen commented:
@swift-ci Please test

kimdv commented:
@swift-ci please test windows

kimdv commented:
@swift-ci please test windows

kimdv commented:
@swift-ci please test Windows

kimdv commented:
@swift-ci please test Windows
